cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
+
+		return $version;
+	}
+
 	/**
 	 * Get the current sync signature of an asset.
 	 *
-	 * @param int $post_id The post ID.
+	 * @param int  $attachment_id The attachment ID.
+	 * @param bool $cached        Flag to specify if a cached signature is to be used or build a new one.
 	 *
 	 * @return array|bool
 	 */
-	public function get_signature( $post_id ) {
+	public function get_signature( $attachment_id, $cached = true ) {
 		static $signatures = array(); // Cache signatures already fetched.
 
-		$return = false;
-		if ( ! empty( $signatures[ $post_id ] ) ) {
-			$return = $signatures[ $post_id ];
+		$return = array();
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cached ) {
+			$return = $signatures[ $attachment_id ];
 		} else {
-			$signature = $this->plugin->components['media']->get_post_meta( $post_id, self::META_KEYS['signature'], true );
-			if ( ! empty( $signature ) ) {
-				$base_signatures        = $this->generate_signature( $post_id );
-				$signatures[ $post_id ] = wp_parse_args( $signature, $base_signatures );
-				$return                 = $signatures[ $post_id ];
+			$signature = $this->managers['media']->get_post_meta( $attachment_id, self::META_KEYS['signature'], true );
+			if ( empty( $signature ) ) {
+				$signature = array();
 			}
+
+			// Remove any old or outdated signature items. against the expected.
+			$signature                    = array_intersect_key( $signature, $this->sync_types );
+			$signatures[ $attachment_id ] = $return;
+			$return                       = wp_parse_args( $signature, $this->sync_types );
 		}
 
 		return $return;
 	}
 
+	/**
+	 * Generate a new Public ID for an asset.
+	 *
+	 * @param int $attachment_id The attachment ID for the new public ID.
+	 *
+	 * @return string|null
+	 */
+	public function generate_public_id( $attachment_id ) {
+
+		$cld_folder = $this->managers['media']->get_cloudinary_folder();
+		$file       = get_attached_file( $attachment_id );
+		$file_info  = pathinfo( $file );
+		$public_id  = $cld_folder . $file_info['filename'];
+
+		return ltrim( $public_id, '/' );
+	}
+
+	/**
+	 * Maybe add a suffix to the public ID if it's not unique.
+	 *
+	 * @param string      $public_id     The public ID to maybe add a suffix.
+	 * @param int         $attachment_id The attachment ID.
+	 * @param string|null $suffix        The suffix to maybe add.
+	 *
+	 * @return string The public ID.
+	 */
+	public function add_suffix_maybe( $public_id, $attachment_id, $suffix = null ) {
+
+		// Test if asset exists by calling just the head on the asset url, to prevent API rate limits.
+		$url         = $this->plugin->components['connect']->api->cloudinary_url( $public_id . $suffix );
+		$req         = wp_remote_head( $url, array( 'body' => array( 'rdm' => wp_rand( 100, 999 ) ) ) );
+		$asset_error = strtolower( wp_remote_retrieve_header( $req, 'x-cld-error' ) );
+		$code        = wp_remote_retrieve_response_code( $req );
```

Comment:
Very cool approach here!

Reply:
Thanks :) - Although this will change a little as it was mostly ignored in this round, as I'm working on this part this week.